### PR TITLE
testng-eclipse #243: detect TestNG version before init

### PR DESCRIPTION
**Pull request by missedone**
this to fix issue https://github.com/cbeust/testng-eclipse/issues/243

previously, the testng eclipse plugin detects the runtime testng version by scanning the jar file name starts with 'testng' on the classpath, but the issue is some 3rd party tools/libs like selenium-server-standalone jar contains the testng classes (e.g. selenium-server-standalone-2.46.0.jar contains testng 6.8.5), so obviously the classpath scan approach has flaw in this case.

actually, we don't need pass the testng version to RemoteTestNG before launching the test process, instead, we can let RemoteTestNG to detect the TestNG version at runtime (by evaluating the value of `org.testng.internal.Version.VERSION`)

### CODE REVIEW

**Comment by missedone** [comment]
ping @cbeust @juherr

File: remote/src/main/java/org/testng/remote/RemoteTestNG.java [hunk 1]
```diff
@@ -27,7 +35,63 @@ public static void main(String[] args) throws ParameterException {
...
-        IRemoteTestNG remoteTestNg = ServiceLoaderHelper.getFirst(ra.version).createRemoteTestNG();
+        Version ver = ra.version;
+        if (ver == null) {  // no version specified on cli, detect ourself
+          String strVer = null;
+          try {
+            // use reflection to read org.testng.internal.Version.VERSION for reason of:
+            // 1. bypass the javac compile time constant substitution
+            // 2. org.testng.internal.Version is available since version 6.6
+            @SuppressWarnings("rawtypes")
+            Class clazz = Class.forName("org.testng.internal.Version");
```
Comment (missedone) [inline remote/src/main/java/org/testng/remote/RemoteTestNG.java]: evaluate `org.testng.internal.Version.VERSION` first, with reflection

File: remote/src/main/java/org/testng/remote/RemoteTestNG.java [hunk 2]
```diff
@@ -27,7 +35,63 @@ public static void main(String[] args) throws ParameterException {
...
+            // trim the version to leave digital number only
+            int idx = strVer.indexOf("beta");
+            if (idx > 0) {
+              strVer = strVer.substring(0, idx);
+            }
+            ver = new Version(strVer);
+          } catch (Exception e) {
+            // testng version < 6.6, ClassNotFound: org.testng.internal.Version
+            // parse the MANIFEST.MF of testng jar from classpath
+            final ClassLoader cl = ClassLoader.getSystemClassLoader();
```
Comment (missedone) [inline remote/src/main/java/org/testng/remote/RemoteTestNG.java]: for versions before 6.6 (actually, just for 6.5.1 and 6.5.2) does not contain `org.testng.internal.Version` class, again, with the classpath scan approach

Comment (juherr) [inline remote/src/main/java/org/testng/remote/RemoteTestNG.java]: Reading Manifest seems easier with this solution:

```java
Enumeration<URL> resources = ClassLoader.getSystemClassLoader().getResources("META-INF/MANIFEST.MF");
for (URL resource : resources) {
    try (InputStream stream = resource.openStream()) {
      Manifest manifest = new Manifest(stream);
      // check that this is your manifest and do what you need or get the next one ...
    } catch (IOException E) {
      // handle
    }
}
```

 http://stackoverflow.com/a/1273196/4234729

Comment (missedone) [inline remote/src/main/java/org/testng/remote/RemoteTestNG.java]: good catch, as it's running on runtime, we can utilize the classloader
thanks

File: remote/src/main/java/org/testng/remote/RemoteTestNG.java [hunk 3]
```diff
@@ -46,6 +59,56 @@ public static void main(String[] args) throws ParameterException {
         }
     }
 
+    private static Version getTestNGVersion() {
+      try {
+        // use reflection to read org.testng.internal.Version.VERSION for reason of:
+        // 1. bypass the javac compile time constant substitution
+        // 2. org.testng.internal.Version is available since version 6.6
+        @SuppressWarnings("rawtypes")
+        Class clazz = Class.forName("org.testng.internal.Version");
```
Comment (missedone) [inline remote/src/main/java/org/testng/remote/RemoteTestNG.java]: with reflection to read the value of `org.testng.internal.Version.VERSION`, in this way, we don't need to care where is the testng classes, even they are packaged in a fat jar, as long as they're on the runtime classpath, I can get the version then.

**Comment by cbeust** [comment]
Looks safer than the previous approach, indeed. 

Still a lot of reflection in there, but since it's an improvement, I'm fine merging it.

**Pull request merged**